src/firefly/js/core/MasterSaga.js

```diff
@@ -16,18 +21,99 @@ export function dispatchAddSaga(saga, params={}) {
     flux.process({ type: ADD_SAGA, payload: { saga,params}});
 }
 
+/**
+ * @param {object}   p
+ * @param {string}   p.id       a unique identifier for this watcher
+ * @param {string[]} p.actions  an array of action types to watch
+ * @param {function} p.callback a callback function to handle the action(s).  The triggered action will be passed back as a parameter.
+ */
+export function dispatchAddWatcher({id, actions, callback}) {
+    flux.process({ type: ADD_WATCHER, payload: {id, actions, callback}});
+}
+
+/**
+ * cancel the watcher with the given id.
+ * @param {string} id  a unique identifier of the watcher to cancel
+ */
+export function dispatchCancelWatcher(id) {
+    flux.process({ type: CANCEL_WATCHER, payload: {id}});
+}
+
+/**
+ * Cancel all watchers.  Should only be called during re-init scenarios. 
+ */
+export function dispatchCancelAllWatchers() {
+    flux.process({ type: CANCEL_ALL_WATCHER});
+}
+
 
 /**
  * This saga launches all the predefined Sagas then loops and waits for any ADD_SAGA actions and launches those Segas
  */
 export function* masterSaga() {
-
+    let watchers = {};
 
     // Start a saga from any action
     while (true) {
-        var action= yield take(ADD_SAGA);
-        const {getState, dispatch}= flux.getRedux();
-        const {saga,params}= action.payload;
-        if (typeof saga === 'function') yield fork( saga, params, dispatch, getState);
+        const action= yield take([ADD_SAGA, ADD_WATCHER, CANCEL_WATCHER, CANCEL_ALL_WATCHER]);
+
+        switch (action.type) {
+            case ADD_SAGA: {
+                const {getState, dispatch}= flux.getRedux();
+                const {saga,params}= action.payload;
+                if (typeof saga === 'function') yield fork( saga, params, dispatch, getState);
+                break;
+            }
+            case ADD_WATCHER: {
+                const {getState, dispatch}= flux.getRedux();
+                const {id, actions, callback}= action.payload;
```

Comment:
I think a watcher should be able to take a params object like add saga.  It might not be as important since many watch would probably closures but there is not reason not to have it.